cpp/ql/src/Likely Bugs/RedundantNullCheckParam.ql

```diff
@@ -0,0 +1,59 @@
+/**
+ * @name Redundant null check or missing null check
+ * @description Checking a parameter for nullness in one path,
+ *              and not in another is likely to be a sign that either
+ *              the check can be removed, or added in the other case.
+ * @kind problem
+ * @id cpp/redundant-null-check-param
+ * @problem.severity recommendation
+ * @tags reliability
+ *       security
+ *       external/cwe/cwe-476
+ */
+
+import cpp
+
+predicate blockDominates(Block check, Block access) {
+  check.getLocation().getStartLine() <= access.getLocation().getStartLine() and
+  check.getLocation().getEndLine() >= access.getLocation().getEndLine()
+}
+
+predicate isCheckedInstruction(VariableAccess unchecked, VariableAccess checked) {
+  checked =
+    any(VariableAccess va |
+      va.getTarget() = unchecked.getTarget()
+  ) and
+//Simple test if the first access in this code path is dereferenced
+  not dereferenced(checked) and
+  blockDominates(checked.getEnclosingBlock(), unchecked.getEnclosingBlock())
+}
+
+pragma[noinline]
+predicate candidateResultUnchecked(VariableAccess unchecked) {
+  not isCheckedInstruction(unchecked, _)
+}
+
+pragma[noinline]
```

Comment:
What is this `pragma` useful for?

Reply:
I do not know if it is useful

I copied it from https://github.com/github/codeql/blob/master/cpp/ql/src/Likely%20Bugs/RedundantNullCheckSimple.ql#L52

What do you think of it ?

Reply:
`pragma[noinline]` is used to force evaluation of the predicate outside of its context (that is, the place its used, the place it could be inlined).  It's sometimes used as a workaround when the optimizer is choosing a poor evaluation order, to force certain parts of the computation to be performed first (in effect) in the `pragma[noinline]` predicate.  I wouldn't recommend using it unless you're aware of a performance issue and you've checked it significantly helps (often it will do more harm than good).  Most of the time the optimizer makes a pretty good choice and you shouldn't need to worry about this.

On the subject of performance, I notice that the `candidateResultChecked` predicate will be very large, because nothing associates `param` to its other parameters (so the predicate is a Cartesian product of all computed values of `check`+`eqop` with all values of `param`; happy to explain this further if there's interest).  In any case, consider moving the line `count(param.getAnAssignment()) = 0` into the main query predicate instead.

Reply:
> pragma[noinline] predicate. I wouldn't recommend using it 

Ok removing it

>  In any case, consider moving the line count(param.getAnAssignment()) = 0 into the main query predicate instead.

Seems reasonable indeed